agent/consul/leader_connect_ca.go

```diff
@@ -0,0 +1,1082 @@
+		return nil, fmt.Errorf("error parsing root fingerprint: %v", err)
+	}
+	rootCert, err := connect.ParseCert(pemValue)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing root cert: %v", err)
+	}
+	keyType, keyBits, err := connect.KeyInfoFromCert(rootCert)
+	if err != nil {
+		return nil, fmt.Errorf("error extracting root key info: %v", err)
+	}
+	return &structs.CARoot{
+		ID:                  id,
+		Name:                fmt.Sprintf("%s CA Root Cert", strings.Title(provider)),
+		SerialNumber:        rootCert.SerialNumber.Uint64(),
+		SigningKeyID:        connect.EncodeSigningKeyID(rootCert.SubjectKeyId),
+		ExternalTrustDomain: clusterID,
+		NotBefore:           rootCert.NotBefore,
+		NotAfter:            rootCert.NotAfter,
+		RootCert:            pemValue,
+		PrivateKeyType:      keyType,
+		PrivateKeyBits:      keyBits,
+		Active:              true,
+	}, nil
+}
+
+// getCAProvider is being called while holding caProviderReconfigurationLock
+// which means it must never take that lock itself or call anything that does.
+func (c *CAManager) getCAProvider() (ca.Provider, *structs.CARoot) {
+	retries := 0
+	var result ca.Provider
+	var resultRoot *structs.CARoot
+	for result == nil {
+		c.providerLock.RLock()
+		result = c.provider
+		resultRoot = c.providerRoot
+		c.providerLock.RUnlock()
+
+		// In cases where an agent is started with managed proxies, we may ask
+		// for the provider before establishLeadership completes. If we're the
+		// leader, then wait and get the provider again
+		if result == nil && c.srv.IsLeader() && retries < 10 {
+			retries++
+			time.Sleep(50 * time.Millisecond)
+			continue
+		}
+
+		break
+	}
+
+	return result, resultRoot
+}
+
+// setCAProvider is being called while holding the stateLock
+// which means it must never take that lock itself or call anything that does.
+func (c *CAManager) setCAProvider(newProvider ca.Provider, root *structs.CARoot) {
+	c.providerLock.Lock()
+	c.provider = newProvider
+	c.providerRoot = root
+	c.providerLock.Unlock()
+}
+
+// InitializeCA sets up the CA provider when gaining leadership, either bootstrapping
+// the CA if this is the primary DC or making a remote RPC for intermediate signing
+// if this is a secondary DC.
+func (c *CAManager) InitializeCA() error {
+	// Bail if connect isn't enabled.
+	if !c.srv.config.ConnectEnabled {
+		return nil
+	}
+
+	// Update the state before doing anything else.
+	err := c.setState(CAStateInitializing, true)
+	if err != nil {
+		return err
+	}
+	defer c.setState(CAStateReady, false)
+
+	// Initialize the provider based on the current config.
+	conf, err := c.initializeCAConfig()
+	if err != nil {
+		return err
+	}
+	provider, err := c.srv.createCAProvider(conf)
+	if err != nil {
+		return err
+	}
+
+	c.setCAProvider(provider, nil)
+
+	// If this isn't the primary DC, run the secondary DC routine if the primary has already been upgraded to at least 1.6.0
+	if c.srv.config.PrimaryDatacenter != c.srv.config.Datacenter {
```

Comment:
Maybe it would make this even clearer if we switched this to:
```
if c.srv.config.PrimaryDatacenter == c.srv.config.Datacenter {
	return c.initializeRootCA(provider, conf)
}
// .. all of the secondary code outdented
```